File: Data Structures & Algorithms/climbing-stairs/submission-2.java
class Solution {
    // Time COmplexity: O(N)
    // Space Complextiy: O(1)
    public int climbStairs(int n) {
        if (n == 1) return 1;
        int dp1 = 1;
        int dp2 = 1;
        int current = 0;

        for (int i = 2; i <= n; i++) {
            current = dp1 + dp2;
            dp2 = dp1;
            dp1 = current;
        }

        return current;
    }
}

File: Data Structures & Algorithms/sliding-window-maximum/submission-2.py
# Initial approach:
# Maintain a heap with k capacity
# Time complexity: O(nlogk)
# Space complexity: O(k)


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = deque() # save index
        result = []
        for i, num in enumerate(nums):
            while q and q[0] <= i - k:
                q.popleft()
            

            while q and nums[q[-1]] <= num:
                q.pop()
            
            q.append(i)
            
            if i >= k - 1:
                result.append(nums[q[0]])

        return result


